libs/base/ui/structure/heading/src/heading.component.ts

```diff
@@ -1,47 +1,55 @@
-import { hydrate, ssrShim } from '@spryker-oryx/utilities';
-import { html, LitElement, TemplateResult } from 'lit';
+import { featureVersion, ssrShim } from '@spryker-oryx/utilities';
+import { LitElement, PropertyValueMap, TemplateResult } from 'lit';
 import { property } from 'lit/decorators.js';
-import { HeadingAttributes, HeadingTag } from './heading.model';
-import { headlineStyles } from './styles/base.styles';
+import { html } from 'lit/static-html.js';
+import {
+  HeadingAttributes,
+  HeadingTag,
+  HeadingVisibility,
+} from './heading.model';
+import { headingStyles } from './heading.styles';
+import { headlineStyles } from './styles';
+import { TypographyController } from './typography.controller';
 
 @ssrShim('style')
-@hydrate()
 export class HeadingComponent extends LitElement implements HeadingAttributes {
-  static styles = headlineStyles;
+  static styles = featureVersion >= '1.4' ? headingStyles : headlineStyles;
 
-  @property({ reflect: true }) tag?: HeadingTag;
-  @property({ reflect: true }) as?: HeadingTag | 'hide';
+  protected typographyController = new TypographyController(this);
 
+  @property() tag?: HeadingTag;
+  @property() typography?: HeadingTag | HeadingVisibility;
+  @property() lg?: HeadingTag | HeadingVisibility;
+  @property() md?: HeadingTag | HeadingVisibility;
+  @property() sm?: HeadingTag | HeadingVisibility;
+  @property() maxLines?: number;
+
+  @property({ reflect: true }) as?: HeadingTag;
   @property({ reflect: true, attribute: 'as-lg' }) asLg?:
     | HeadingTag
-    | 'hide'
-    | 'show';
-
+    | HeadingVisibility;
   @property({ reflect: true, attribute: 'as-md' }) asMd?:
     | HeadingTag
-    | 'hide'
-    | 'show';
-
+    | HeadingVisibility;
   @property({ reflect: true, attribute: 'as-sm' }) asSm?:
     | HeadingTag
-    | 'hide'
-    | 'show';
+    | HeadingVisibility;
 
-  @property() set maxLines(value: number) {
-    if (value > 0) {
-      this.style.setProperty('--max-lines', String(value));
-    }
+  protected override render(): TemplateResult {
+    return this.renderTag(html`<slot></slot>`);
   }
 
-  protected override render(): TemplateResult {
-    return html`${this.renderTag(html`<slot></slot>`)}`;
+  protected willUpdate(
+    properties: PropertyValueMap<any> | Map<PropertyKey, unknown>
+  ): void {
+    super.willUpdate(properties);
+    // as long as the controller does not support willUpdate on SSR,
+    // we need to call this manually.
+    this.typographyController.hostUpdate();
```

Comment:
What about wrapping it with `isServer`? as willUpdate will be called multiple times on the client (and calling controller should not be needed in those cases)

Reply:
yes we had that and I forgot to bring it back